libclc/utils/libclc-remangler/LibclcRemangler.cpp

```diff
@@ -147,340 +123,606 @@ class BumpPointerAllocator {
+        DeclarationNameInfo(), VoidFuncType,
+        ContextAST->getTrivialTypeSourceInfo(ContextAST->VoidTy), SC_None,
+        false, false, false, ConstexprSpecKind::Unspecified, nullptr);
+    FDSpecialization->setImplicitlyInline(false);
+
+    FDSpecialization->setDeclName(&ContextAST->Idents.get(KernelName));
+
+    // Used to populate all the template type param decls into.
+    NamespaceDecl *DummyNamespace = NamespaceDecl::Create(
+        *ContextAST, ContextAST->getTranslationUnitDecl(), false,
+        SourceLocation(), SourceLocation(),
+        &ContextAST->Idents.get("DummyNamespace", tok::TokenKind::identifier),
+        nullptr);
+    DummyNamespace->setImplicit(true);
+    ContextAST->getTranslationUnitDecl()->addDecl(DummyNamespace);
+
+    // Will be used to build template parameter list.
+    SmallVector<NamedDecl *> TemplateNamedDecls;
+    // Used for setting template specialisation.
+    SmallVector<TemplateArgument> TemplateArguments;
+    // Used for the fixups.
+    SmallVector<clang::QualType> TemplateTypeParamTys;
+    unsigned TemplateIndex = 0;
+    for (auto &TemplateArgQT : TemplateArgTys) {
+      std::string const Name{std::string{"TempTy"} +
+                             std::to_string(TemplateIndex)};
+      auto &II = ContextAST->Idents.get(Name);
+      auto *TTPD = TemplateTypeParmDecl::Create(
+          *ContextAST, DummyNamespace, SourceLocation(), SourceLocation(), 0,
+          TemplateIndex, &II, /* Typenaem */ true, /*ParameterPack*/ false);
+      TTPD->setDefaultArgument(
+          ContextAST->getTrivialTypeSourceInfo(TemplateArgQT));
+
+      TemplateNamedDecls.emplace_back(TTPD);
+      auto TA = TemplateArgument(TemplateArgQT);
+      TemplateArguments.emplace_back(TA);
+
+      // Store this qualified type with newly created proper template type
+      // param qualified type.
+      TemplateTypeParamTys.push_back(
+          ContextAST->getTemplateTypeParmType(0, TemplateIndex, false, TTPD));
+
+      ++TemplateIndex;
+    }
+    // Fix up the template types in the original FD's arg tys and return ty.
+    auto AreQTsEqual = [&](clang::QualType &LHS, clang::QualType &RHS) -> bool {
+      return (LHS.getBaseTypeIdentifier() && RHS.getBaseTypeIdentifier() &&
+              LHS.getBaseTypeIdentifier()->isStr(
+                  RHS.getBaseTypeIdentifier()->getName())) ||
+             LHS == RHS;
+    };
+    unsigned NumReplaced = 0;
+    unsigned Idx = 0;
+    for (auto &TemplateArgQT : TemplateArgTys) {
+      if (AreQTsEqual(TemplateArgQT, RetTy)) {
+        RetTy = TemplateTypeParamTys[Idx];
+        goto Found;
+      }
+      for (unsigned i = 0; i < ArgTys.size(); ++i) {
+        if (AreQTsEqual(ArgTys[i], TemplateArgQT)) {
+          ArgTys[i] = TemplateTypeParamTys[Idx];
+          goto Found;
+        }
+      }
+    Found:
+      ++NumReplaced;
+      ++Idx;
+    }
+    assert(NumReplaced >= TemplateTypeParamTys.size() &&
+           "Expected full specialisation.");

```

Comment:
Use American spelling to be consistent?
```suggestion
           "Expected full specialization.");
```